include/rocksdb/sst_partitioner.h

```diff
@@ -0,0 +1,81 @@
+// Copyright (c) 2011-present, Facebook, Inc.  All rights reserved.
+//  This source code is licensed under both the GPLv2 (found in the
+//  COPYING file in the root directory) and Apache 2.0 License
+//  (found in the LICENSE.Apache file in the root directory).
+// Copyright (c) 2012 The LevelDB Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file. See the AUTHORS file for names of contributors.
+//
+// Class for specifying user-defined functions which perform a
+// transformation on a slice.  It is not required that every slice
+// belong to the domain and/or range of a function.  Subclasses should
+// define InDomain and InRange to determine which slices are in either
+// of these sets respectively.
+
+#pragma once
+
+#include <memory>
+#include <string>
+
+#include "rocksdb/rocksdb_namespace.h"
+#include "rocksdb/slice.h"
+
+namespace ROCKSDB_NAMESPACE {
+
+class Slice;
+
+/*
+ * A SstPartitioner is a generic pluggable way of defining the partition
+ * of SST files. Compaction job will split the SST files on partition boundary
+ * to lower the write amplification during SST file promote to higher level.
+ */
+class SstPartitioner {
+ public:
+  virtual ~SstPartitioner(){};
+
+  // Return the name of this partitioner.
+  virtual const char* Name() const = 0;
+
+  // Returns true if partition boundary was detected and compaction should
+  // create new file.
+  virtual bool ShouldPartition(const Slice& last_user_key,
```

Comment:
Shall we create a new options struct for `ShouldPartition` parameter, so that the interface is extendable?

Reply:
I am not sure I understand. We have the Context there already. If you want to pass struct then we should place there also prev and current key there. Is that what you mean? I have also fixed documentation to make the purpose easier to understand.

Reply:
Yes, that's what I mean. It makes the API easier to extend later.

Reply:
I have changed to   
`virtual PartitionerResult ShouldPartition(const PartitionerRequest& request) = 0;`